lib-src/rpc2/rpc2-src/pack_helper.c

```diff
@@ -0,0 +1,383 @@
+		/* if (mode != NO_MODE && who == RP2_CLIENT) fputc('*', where); */
+        *ptr = (RPC2_String)(buf->buffer);
+    }
+    buf->buffer += _PAD(length);
+    return 0;
+}
+
+
+int unpack_countedbs(BUFFER *buf, unsigned char **ptr, RPC2_Unsigned *len_ptr,
+        int who)
+{
+    if (buf->buffer + 4 > buf->eob)
+        return -1;
+    if (who == STUBSERVER) {
+		/* Special hack */
+        *len_ptr = ntohl(*(RPC2_Integer *)(buf->buffer));
+        (buf->buffer) += 4;
+        if (buf->buffer + _PAD(*len_ptr) > buf->eob)
+            return -1;
+        *ptr = (RPC2_Byte *)(buf->buffer);
+        buf->buffer += _PAD(*len_ptr);
+        return 0;
+    } else {
+        *len_ptr = ntohl(*(RPC2_Integer*)(buf->buffer));
+        buf->buffer += 4;
+        if (buf->buffer + _PAD(*len_ptr) > buf->eob)
+            return -1;
+		/*    bug fix. Should update SeqLen and use select. M.K. */
+		/*   fprintf(where, "
+		    memcpy((char *)%s->SeqBody, %s, (int32_t)%s);\n", */
+        memcpy(*ptr, buf->buffer, *len_ptr);
+		/*				inc(ptr, length, where); */
+        buf->buffer += _PAD(*len_ptr);
+        return 0;
+    }
+}
+
+
+int unpack_boundedbs(BUFFER *buf, unsigned char **ptr, RPC2_Unsigned *len_ptr,
+        RPC2_Unsigned *max_len_ptr, int who, int mode)
+{
+    if (buf->buffer + 8 > buf->eob)
+        return -1;
+    if (who == STUBSERVER && mode != STUBIN) {
+        *max_len_ptr = ntohl(*(RPC2_Unsigned *)(buf->buffer));
+    }
+    buf->buffer += 4; /* Skip maximum length */
+    if ((who == STUBCLIENT && mode != STUBIN) ||
+            (who == STUBSERVER && mode != STUBOUT)) {
+        *len_ptr = ntohl(*(RPC2_Unsigned *)(buf->buffer));
+    } else if (who == STUBSERVER)
+        *len_ptr = 0;
+    buf->buffer += 4; /* skip packed sequence length */
+    if (who == STUBSERVER && mode == STUBIN)
+        *max_len_ptr = *len_ptr;
+    else {
+        if (*len_ptr > *max_len_ptr)
+            return -1;
+    }
+
+    if (buf->buffer + _PAD(*len_ptr) > buf->eob)
+        return -1;
+    if (who == STUBCLIENT) {
+        if (mode != STUBIN) {
+            memcpy((*ptr), buf->buffer, *len_ptr);
+            buf->buffer += _PAD(*len_ptr);
+        }
+    } else {
+        if (*max_len_ptr != 0) {
+            *ptr = (RPC2_String)malloc(*max_len_ptr);
+            if (*ptr == NULL)
+                return -1;
+            memcpy((*ptr), buf->buffer, *len_ptr);
+            buf->buffer += _PAD(*len_ptr);
+        } else {
+            *ptr = NULL;
+        }
+    }
+    return 0;
+
+}
+
+
+int unpack_encryptionKey(BUFFER *buf, char *ptr)
+{
+    if (buf->buffer + _PAD(RPC2_KEYSIZE) > buf->eob)
+        return -1;
+    memcpy(ptr, buf->buffer, RPC2_KEYSIZE);
+    buf->buffer += _PAD(RPC2_KEYSIZE);
+    return 0;
+}
+
+
+int pack_int(BUFFER *buf,  RPC2_Integer value)
+{
+    if (buf->buffer + 4 > buf->eob)
+        return -1;
+    *(RPC2_Integer *)(buf->buffer) = htonl(value);
+    buf->buffer += 4;
+    return 0;
+}
```

Comment:
Can we coopt the pack functions to also double as a length check function. We can set `buf->buffer = buf->eob = NULL`, then change the functions to do
```
  int pack_int(BUFFER *buf, RPC2_Integer value)
  {
    if (buf->eob) {
      if (buf->buffer + 4 > buf->eob)
        return -1;
      *(RPC2_Integer *)(buf->buffer) = htonl(value);
    }
    buf->buffer += 4;
    return 0;
  }
```
This way the overflow checks and data copies are only done when EOB is specified, and we end up with (intptr_t)buf->buffer containing the length of what would have been packed.